components/editor/utils/keyHandlers.js

```diff
@@ -0,0 +1,155 @@
+import { Block } from 'slate'
+import { matchBlock } from './'
+import { getClosestInSelection } from './selection'
+
+const focusNext = change => {
+  const { value } = change
+  const nextBlock = value.document.getNextBlock(value.endBlock.key)
+  if (nextBlock) {
+    return change.collapseToStartOf(
+      nextBlock
+    )
+  }
+  return true
+}
+
+const focusPrevious = change => {
+  const { value } = change
+  const nextBlock = value.document.getPreviousBlock(value.endBlock.key)
+  if (nextBlock) {
+    return change.collapseToEndOf(
+      nextBlock
+    )
+  }
+  return true
+}
+
+const insertAfter = (change, afterType, insertAfterType) => {
+  const { value } = change
+  const rootNode = (
+      insertAfterType &&
+      value.document.getFurthest(
+        value.endBlock.key, matchBlock(insertAfterType)
+      )
+    ) ||
+    value.document
+
+  const index = rootNode.nodes.findIndex(
+    n => !!n.findDescendant(m => m.key === value.endBlock.key)
+  )
+
+  if (index !== -1) {
+    return change.insertNodeByKey(
+      rootNode.key,
+      index + 1,
+      Block.create({ type: afterType })
+    )
+  }
+  return change
+}
+
+export const staticKeyHandler = ({ TYPE, rule }) => {
+  const { afterType, insertAfterType } = rule.editorOptions
+
+  return (event, change) => {
+    const isBackspace = event.key === 'Backspace'
+    const isEnter = event.key === 'Enter'
+    const isTab = event.key === 'Tab'
+    const isDelete = event.key === 'Delete'
+
+    if (!isBackspace && !isEnter && !isDelete && !isTab) return
+    const { value } = change
+    const inSelection = value.blocks.some(matchBlock(TYPE))
+
+    if (inSelection && value.isExpanded) {
+      if (isBackspace || isDelete) {
+        return change.collapseToStart()
+      }
+      return change.collapseToEnd()
+    }
+
+    const previousBlock = value.document.getPreviousBlock(
+      value.startBlock.key
+    )
+    const isCollapsed = value.isCollapsed
+    const cursorAtStart = isCollapsed &&
+    value.selection.hasStartAtStartOf(value.startBlock)
+
+    if (
+        !inSelection && (
+        !matchBlock(TYPE)(previousBlock) ||
+        !cursorAtStart ||
+        !isBackspace
+      )
+    ) return
+
+    const nextBlock = value.document.getNextBlock(value.endBlock.key)
+    if (isTab) {
+      if (nextBlock) {
+        event.preventDefault()
+        return focusNext(change)
+      }
+      return
+    }
+
+    if (isEnter) {
+      event.preventDefault()
```

Comment:
missing support for soft break with shift: `change.insertText('\n')`

since the lead becomes a static one this is mandatory—we need soft break support in leads

Reply:
This will have to be exposed in the template, I guess. While soft breaks in leads are ok, in figure captions et al they're certainly not.
Will do.

Reply:
already supported now and exposed in template. why not in figure captions?

Reply:
Because it looks ugly? Where is this exposed in template?
I mean, we could turn that for all static fields by default, no problem.

Reply:
yes please turn on by default to match status quo. Br is part of the `globalInlines` in the template.

Reply:
Will do.
You mean its supported through the presence of a certain serializer. But I really wouldn't want to loop through the child serializers to determine whether the UI should capture soft breaks or not.

Reply:
Agree. No need to loop through—just do it globally. All module that start to use `staticKeyHandler` had the support for soft break before expect for `figureCaption` but that was an oversight when reviewing. I agree that it shouldn't be used often in figure captions but there are edge cases that it can be useful for.